JDK 17 CI nightly job

johanandren opened the pull request:
References #31079 

Adds a JDK 17 nightly job along with some fixes for tests to pass.

johanandren committed:
First try JDK 17 CI nightly

patriknw commented on .github/workflows/nightly-builds.yml:
```diff
@@ -120,12 +120,14 @@ jobs:
         # full version from it.
         scalaVersion: ["2.12", "2.13"]
-        jdkVersion: ["adopt@1.8.0", "adopt@1.11"]
+        jdkVersion: ["adopt@1.8.0", "adopt@1.11", "openjdk@1.17.0"]
```
I thought we had to open up some stuff?

johanandren replied:
I think so too, but wanted to see Java tell me that to be sure.

johanandren replied:
(I've been building on JDK 17 locally for a while but I guess the open up part will be needed to actually run anything)

johanandren replied:
Minimal open-ups added to the build now.

johanandren committed:
Allow manually triggering nightlies?

johanandren committed:
Better yaml syntax?

johanandren commented:
Summary of test failures on JDK 17:

* akka-actor-tests:
  * akka.actor.CoordinatedShutdownSpec - reflection in test: Unable to make field private static java.util.IdentityHashMap java.lang.ApplicationShutdownHooks.hooks accessible: module java.base does not "opens java.lang” 
  * akka.dispatch.ForkJoinPoolStarvationSpec - consistently fails not sure why, no sign of internal access

* akka-cluster-sharding
    * akka.cluster.sharding.DDataClusterShardingRememberEntitiesEnabled - Unable to make field long java.nio.Buffer.address accessible: module java.base does not "opens java.nio"

* akka-remote
    * akka.remote.classic.transport.netty.NettyTransportSpec - ah bind to 127.0.1.1 which isn’t set on MacOS
    * akka.remote.classic.ActorsLeakSpec - consistently fails locally, no sign it is JDK internal API-related
    * akka.remote.ArteryMessageLoggingSpec - Unable to make field private final java.util.Set sun.nio.ch.SelectorImpl.selectedKeys accessible: module java.base does not "opens sun.nio.ch" to unnamed module
    * akka.remote.artery.ArteryUpdSendConsistencyWithOneLaneSpec - Unable to make field private final java.util.Set sun.nio.ch.SelectorImpl.selectedKeys accessible: module java.base does not "opens sun.nio.ch" to unnamed module
    * akka.remote.artery.BindCanonicalAddressSpec - Unable to make field private final java.util.Set sun.nio.ch.SelectorImpl.selectedKeys accessible: module java.base does not "opens sun.nio.ch" to unnamed module
    * akka.remote.artery.aeron.AeronSinkSpec - Unable to make field private final java.util.Set sun.nio.ch.SelectorImpl.selectedKeys accessible: module java.base does not "opens sun.nio.ch" to unnamed module
    * akka.remote.artery.ArteryUpdSendConsistencyWithThreeLanesSpec - Unable to make field private final java.util.Set sun.nio.ch.SelectorImpl.selectedKeys accessible: module java.base does not "opens sun.nio.ch" to unnamed module

*  akka-distributed-data
    * akka.cluster.ddata.DurableDataSpec - Unable to make field long java.nio.Buffer.address accessible: module java.base does not "opens java.nio" to unnamed module
    * akka.cluster.ddata.DurableDataWriteBehindSpec - Unable to make field long java.nio.Buffer.address accessible: module java.base does not "opens java.nio" to unnamed module
    * akka.cluster.ddata.protobuf.ReplicatedDataSerializerSpec - not sure what this is about - serialization formats consistently differ


Looks like we need to open up (and maybe document opening up) `java.nio` across the board for now, I'll investigate the other failures and see if we can avoid reflecting into java.lang in the coordinated shutdown test (annoying to open up `java.lang` just for that test case)

johanandren committed:
Add "opens" for nio internals aeron and leveldb uses

johanandren committed:
Avoid reflecting into JDK internals to check Coordinated shutdown JVM hooks

rather than adding a blanket java.lang opens

johanandren commented:
There are two tests that still consistently fails on my machine on JDK 17, that I don't understand:
`akka.remote.classic.ActorsLeakSpec` and `akka.dispatch.ForkJoinPoolStarvationSpec`

Perhaps those are just racy on a faster machine (but neither fails on JDK 11), not sure. Would be good to let CI run and see if they fail there as well.

johanandren committed:
Skip old wire format compatibility check for ddata ORMaps

Doesn't pass for JDK 17, but change was introduced in 2.5.0, no need
to promise distributed data wire compatibility with 2.4.x

johanandren commented on project/JdkOptions.scala:
```diff
@@ -26,6 +26,16 @@ object JdkOptions extends AutoPlugin {
+      "--add-opens=java.base/sun.nio.ch=ALL-UNNAMED" ::
+        // for LevelDB
+        "--add-opens=java.base/java.nio=ALL-UNNAMED" :: Nil
+    } else Nil
```
Added here rather than CI build config to make it buildable/testable locally on JDK 17 and always passed to the multi jvm tests when on JDK 17 and later.

ennru replied:
The code says `isJdk11orHigher`?

johanandren replied:
ups, should ofc be isJdk17OrHigher

johanandren committed:
sbt file formatting

johanandren committed:
Wrong condition in jdkOptions

patriknw commented on project/JdkOptions.scala:
```diff
@@ -26,6 +26,16 @@ object JdkOptions extends AutoPlugin {
+  val versionSpecificJavaOptions =
+    if (isJdk17orHigher) {
+      // for aeron
+      "--add-opens=java.base/sun.nio.ch=ALL-UNNAMED" ::
```
we should add a note about this in the docs

johanandren replied:
I created an issue for that already #31087

patriknw reviewed (approved):
LGTM with suggestion of doc follow up

raboof commented on .github/workflows/nightly-builds.yml:
```diff
@@ -3,6 +3,7 @@ name: Nightly Builds
 on:
   schedule:
     - cron: "0 0 * * *"
+  workflow_dispatch:
```
Do I understand correctly that this means you *can* trigger this workflow manually or from another workflow, but we're not doing that anywhere automatically yet?

johanandren replied:
Yes, mostly to be able to trigger-test changes manually, not so much for the trigger-from-another-workflow aspect

raboof reviewed (approved):
Agreed we should add a note about the required `--add-opens=java.base/sun.nio.ch=ALL-UNNAMED` (for aeron) and `--add-opens=java.base/java.nio=ALL-UNNAMED` (for leveldb) to the docs, but that can be a follow-up PR

johanandren commented:
Also fixes/handles #30341 #30339
